### DISCUSSION TITLE
Am I able to add a REST API and caching layer that sits behind Kong?

### DISCUSSION THREAD

**Discussion by gabedealmeida** [supabase/supabase#1994] [Questions] [answered]
Hey Supabase Team!

I'm interested in using Supabase for an iOS app that I'll be building. Instead of interacting with the backend through the client library directly from the app, I'm interested in using a REST API so that if I want to change how I do things on the backend after the app is launched, I can do so immediately on the REST API. I was wondering if there's an easy way to build an API that sits behind Kong to take advantage of all that Kong provides.

The flow would be something like ex:
* User wants to log in
* User first makes a request to the API endpoint that I build ex, blahblah.com/v1/my-app/login or blahblah.supabase.io/v1/my-app/login
* Kong makes sure that the user is authenticated with the anon API key
* Request reaches my API
* Inside of my API, I make a request to the Gotrue API to log the user in and get the JWT token. Perhaps there's an easy way to do this that doesn't require hitting Kong up again, i.e., service-to-service communication
* The token is returned to the user

Now two weeks after the app is launched, I decide that I should also increment their points on the daily-login table when a user is logging in. So all I have to do is make the change on my API; instead of having to go on XCode and put all this new logic there, have to release a new version of the app and hope most users update their app. 

Another example:
* User wants to fetch images
* User first makes a request to the API endpoint that I build ex, blahblah.com/v1/my-app/images or blahblah.supabase.io/v1/my-app/images
* Kong makes sure that the user is authenticated, also taking advantage of all the Gotrue logic that is necessary if need be
* Request reaches my API
* Images are returned to the user

Additionally, I was wondering if I'm able to introduce a caching layer in the architecture. A Redis instance/cluster that sits behind Kong and that can be accessed from my REST API to lessen reads on the DB.

Thank you all so very much!

Very best,
Gabriel

**Comment by steve-chavez** [comment, marked as answer]
Hey @gabedealmeida,

You can freely connect our hosted database to any other components you might have, but our middleware(that includes Kong) is not open in the hosted version - it would be too hard for us to maintain quality of service if we'd let third party components in, at least for now.

So one option would be to host your own middleware that connects to our hosted database, or you can use the self-hosted version [supabase-cli](https://github.com/supabase/cli) were you can control every component.
